rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,777 @@
+generation, signing operations, and signature verification.  Some of the more
+restrictive security policies are starting to reject RSA2048 (e.g.
+[RHEL 8's FUTURE policy](https://access.redhat.com/articles/3642912)).
+
+## Details
+
+### Summary
+
+We will introduce the concept of signature algorithm "suites" that will
+configure the key types and signature algorithms used across Teleport.
+This will be a cluster-wide configuration available in both the `teleport.yaml`
+config of Teleport Auth servers (so that initial CA keys generated at first
+startup can be configured) and in the `cluster_auth_preference` (so that Cloud
+users can optionally change the suite used in their cluster).
+
+The initial set of suites we will introduce is:
+
+* `legacy`
+* `balanced-v1`
+* `fips-v1`
+* `hsm-v1`
+
+The `legacy` suite exists for compatibility and it will use the exact same
+key types and algorithms Teleport uses today (2048-bit RSA everywhere).
+This will be the initial default, changing the default to `balanced-v1` will
+be a breaking change that will be made in either a major version release of
+Teleport, or with a version bump to the `teleport.yaml` config.
+
+The `balanced-v1` suite will use a modern set of algorithms selected to
+balance security, compatibility, and performance.
+The proposed selection for this suite is to use Ed25519 as the default choice,
+and ECDSA with the NIST P-256 curve in cases where Teleport is likely to
+interact with third-party software that does not support Ed25519.
+The exception is the database CA, where it looks like it will be necessary to
+continue using RSA in order to continue supporting our current Snowflake
+integration.
+
+The `fips-v1` suite will only use key types and signature algorithms that are
+approved by FIPS 186-5 *and* are supported in Go's `GOEXPERIMENT=boringcrypto`
+mode.
+FIPS 186-5 actually approves all algorithms in the `balanced-v1` suite (and
+even the `legacy` suite), but Go's BoringCrypto integration does NOT support
+Ed25519.
+This suite will be based off of the `balanced-v1`, but all uses of Ed25519 will
+be replaced by ECDSA with the NIST P-256 curve.
+Teleport will fail to start in FIPS mode unless the `fips-v1` or `legacy` suite
+is selected.
+The `fips-v1` algorithm suite will be available without FIPS mode, to enable
+migration to FIPS mode.
+
+The `hsm-v1` suite will be based off of the `balanced-v1` suite, but all uses of
+Ed25519 *for CA keys only* will be replaced by ECDSA with the NIST P-256 curve.
+This is necessary and sufficient for all of the HSMs and KMSs we currently
+support and test for (YubiHSM2, AWS CloudHSM, and GCP KMS).
+This is very similar to the `fips-v1` suite, but "subjects" (users, hosts, etc)
+will still be allowed to use Ed25519 keys, it is only the CAs that are limited
+to the algorithms supported by the HSM.
+Whenever the default suite is updated from "legacy" to "balanced-v1", it will be
+overridden by `fips-v1` if and HSM or KMS is configured.
```

Comment:
If we are overbidding the cipher suite from other configuration values, will we log about the conflict?

Reply:
I should reword this. I mean that if the suite is not explicitly set in any config, we will have to select a default. Initially this default will probably be `legacy`. But whenever we update the default, it should be `balanced-v1` if no HSM or KMS is configured, and `hsm-v1` if there is an HSM or KMS configured. So it's not really a conflict, just a default selection. We'll log it anyways though, probably just at debug or info level.

If they have an HSM and they explicitly configure `balanced-v1`, we can warn but try to support it in case their PKCS#11 HSM supports Ed25519, and fail early with explicit logs if there's an error